Source/MQTTnet/Server/Internal/MqttClientSubscriptionsManager.cs

```diff
@@ -60,13 +60,16 @@ public CheckSubscriptionsResult CheckSubscriptions(string topic, ulong topicHash
 
                 foreach (var wcs in _wildcardSubscriptionsByTopicHash)
                 {
-                    var wildcardSubscriptions = wcs.Value;
                     var subscriptionHash = wcs.Key;
-                    var subscriptionHashMask = wildcardSubscriptions.HashMask;
-
-                    if ((topicHash & subscriptionHashMask) == subscriptionHash)
+                    var subscriptionsByHashMask = wcs.Value.SubscriptionsByHashMask;
+                    foreach (var shm in subscriptionsByHashMask)
                     {
-                        possibleSubscriptions.AddRange(wildcardSubscriptions.Subscriptions.ToList());
+                        var subscriptionHashMask = shm.Key;
+                        if ((topicHash & subscriptionHashMask) == subscriptionHash)
+                        {
+                            var subscriptions = shm.Value;
+                            possibleSubscriptions.AddRange(subscriptions.ToList());
```

Comment:
Is the call of "ToList()" really required? If not, it is only a useless allocation.

Reply:
Good point, that's redundant. I did not notice a change in performance with a cursory look; maybe it's optimized away. I'll do some benchmarking to compare against https://github.com/dotnet/MQTTnet/pull/1298#issuecomment-1036967766 to see if anything else has changed, also in respect of the new AsyncLock.